cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBMessageTable.java

```diff
@@ -0,0 +1,204 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.messaging.store.leveldb;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.messaging.MessagingUtils;
+import co.cask.cdap.messaging.data.MessageId;
+import co.cask.cdap.messaging.store.MessageTable;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.tephra.Transaction;
+import org.iq80.leveldb.DB;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * LevelDB implementation of {@link MessageTable}.
+ */
+public class LevelDBMessageTable implements MessageTable {
+  private static final Gson GSON = new Gson();
+  private static final Type STRING_MAP_TYPE = new TypeToken<Map<String, byte[]>>() { }.getType();
+  private static final String PAYLOAD_COL = "p";
+  private static final String TX_COL = "t";
+  private static final byte[] SEPARATOR = new byte[] { ':' };
+
+  private final DB levelDB;
+  private volatile long writeTimestamp;
+  private volatile short pSeqId;
+
+  public LevelDBMessageTable(DB levelDB) {
+    this.levelDB = levelDB;
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, long startTime, int limit, @Nullable Transaction transaction)
+    throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(startTime));
+    byte[] stopKey = Bytes.stopKeyForPrefix(topic);
+    return new LevelDBCloseableIterator(levelDB, startRow, stopKey, limit, transaction);
+  }
+
+  @Override
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive, final int limit,
+                                        @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(messageId.getPublishTimestamp()),
+                                Bytes.toBytes(messageId.getSequenceId()));
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    byte[] stopKey = Bytes.stopKeyForPrefix(topic);
+    return new LevelDBCloseableIterator(levelDB, startRow, stopKey, limit, transaction);
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    long writeTs = System.currentTimeMillis();
```

Comment:
Seems like we can have a common parent between levelDb and HBase implementation? Many of the logic are shared. LevelDB also has batchWrite (put and delete) which helps refactoring easier.